### ISSUE TITLE
Multi-Platform support

### ISSUE THREAD

**Issue by jhjacobs81** [linuxkettle/AppShop#1]
Could this be done?

i actually like the name appShop over helloShop because it removes “the connection to” helloSystem. 

A multiplatform “appstore” would be a fantastic way for dev’s to promote their software.. 

We could make guidelines for each platform? Maybe say “apps need to be written in this and this language” and then when a user installs the appshop binaries we could make sure the dependencies include the programming lnguage etc?

**Comment by kettle-7** [comment]
It could definitely be done. Both python and qt are cross platform (although I don't believe the latter works on MS-DOS or Windows ME or earlier)

The issue would be figuring out how to get a proper C compiler onto Windows, and at this point I might make 3 backends for the three major platforms, because Windows and macOS can't run ELF binaries.

I ported KolourPaint, it's in apps/Graphics/KolourPaint.tar.gz :~)

**Comment by jhjacobs81** [comment]
Maybe focus on Linux/BSD first ;-) 

i had hopes for hS, but it seems we are too different for it to work (for me) so i’ll keep looking for a distro that share the same goals as me (the beauty of Apple with the privacy of linux and the stability of bsd)

**Comment by kettle-7** [comment]
You could try elementary, the latest version is very good and it uses a 3-year-old base that's well tested and stable. And it's absolutely beautiful, and has an app centre :~)

**Comment by jhjacobs81** [comment]
Elementary looks like a cartoon ;-) 

but yeah, ive tried it ;-)

**Comment by jhjacobs81** [comment]
I would imagine, if all apps are written in Vala or Python.. would this provide a platform native GUI?? 

If we could then ask all programmers to adhere to certain guidelines, keep the wuality of apps to a certain level, people are willing to pay indie dev’s i am sure of this. appshop could create a backend for items on startup, keep a list of files to be removed when the app is removed. Imagine we go with the hS idea of app bundles.. we could run a daemon that keeps track of files to remove if the programs report these files to the correct daemon? Shouldn’t be too hard to write into a program. 

Then if all apps go into /Application or ~/Application this too could become a standard. 

Eventually, appshop could make sure vala and python are installed, so programmers need not wory about all of this, and just focus on their program?

**Comment by kettle-7** [comment]
> if all apps are written in Vala or Python.. would this provide a platform native GUI??

Vala, in a word: NO. Vala is very closely paired with the entire GNOME stack, to the point where it works on Linux but not Windows or macOS, and is unwelcome on FreeBSD (whether helloSystem or not).

Python: Qt does a good job at looking native, tkinter does better but that's a mess.

Java is another good one with a perk over any other compiled language: The same Java app, compiled and all, is one file like an AppImage, but works exactly the same on all platforms.

> Then if all apps go into /Application or ~/Application this too could become a standard.

Well, if we've got desktop2app, who says we can't have an app2desktop?

> If we could then ask all programmers to adhere to certain guidelines, keep the wuality of apps to a certain level, people are willing to pay indie dev’s i am sure of this. appshop could create a backend for items on startup, keep a list of files to be removed when the app is removed. Imagine we go with the hS idea of app bundles.. we could run a daemon that keeps track of files to remove if the programs report these files to the correct daemon? Shouldn’t be too hard to write into a program.

This is going a bit further than I originally planned, but OK!

**Comment by kettle-7** [comment]
If you have a little look at the QML code, and scroll past all of the data-managing stuff that I'm messing about with, you'll probably notice why I like QML so much: simplicity.

**Comment by kettle-7** [comment]
And this is what the data-grabbing garbage gave me:
```js
[{"Category":"Graphics & Film","Tree":{"name":"<anonymous>","subtrees":[],"apps":[{"name":"<anonymous>","subtrees":[],"apps":[],"value":"<empty>","extra":[],"fields":[["Name","KolourPaint"],["Icon","KolourPaint.png"],["Version","20.12.3"],["Summary","KolourPaint is a simple painting program to quickly create raster images. It is useful as a touch-up tool and simple image editing tasks."],["Description","<img src=\"https://kde.org/images/screenshots/kolourpaint.png\" alt=\"KolourPaint\"/><p>KolourPaint is a simple painting program to quickly create raster images. It is useful as a touch-up tool and simple image editing tasks.</p>\n<h3>Features:</h3>\n<ul>\n\t<li>Support for drawing various shapes - lines, rectangles, rounded rectangles, ovals and polygons</li>\n\t<li>Curves, lines and text</li>\n\t<li>Colour picker</li>\n\t<li>Selections</li>\n\t<li>Rotation, monochrome and other advanced effects</li>\n</ul>\n\n<p>For more information see the <a href=\"http://www.kolourpaint.org/\">project website</a>.\n"],["Filename","KolourPaint.tar.gz"],["Gradient0","blue"],["Gradient1","red"],["Foreground","white"]],"path":"https://raw.githubusercontent.com/linuxkettle/AppShop/main/apps/Graphics"}],"value":"<empty>","extra":[],"fields":[],"path":"https://raw.githubusercontent.com/linuxkettle/AppShop/main/apps/Graphics","category":"Graphics & Film"}}]
```
Data garbage without the grabbing part.

**Comment by jhjacobs81** [comment]
Well i'm just the guy with a vision ;-) I have no knowledge of the specific details other then what i read in my daily RSS feeds ;-) All i know is that, what Apple does with its platform, the FOSS world could do too! if only they cared enough to unite :( 

But to this day, most *X distro's look too cartoonish if you come from macOS, give me a window manager thats equally beautiful and ill switch :) coupled with a good appstore with quality apps ofcourse..

**Comment by kettle-7** [comment]
> But to this day, most *X distro's look too cartoonish if you come from macOS, give me a window manager thats equally beautiful and ill switch :) coupled with a good appstore with quality apps ofcourse..

That's the sad truth. People try to make *nix look like something it isn't, and make it look cartoonish. Then some people (cough cough gnome) absolutely go too far the opposite way and make an absolute abomination. I must admit, GNOME 40 is good for productivity, but bad for many other things.

**Comment by jhjacobs81** [comment]
> > if all apps are written in Vala or Python.. would this provide a platform native GUI??
> 
> Vala, in a word: NO. Vala is very closely paired with the entire GNOME stack, to the point where it works on Linux but not Windows or macOS, and is unwelcome on FreeBSD (whether helloSystem or not).
> 
> Python: Qt does a good job at looking native, tkinter does better but that's a mess.
> 
> Java is another good one with a perk over any other compiled language: The same Java app, compiled and all, is one file like an AppImage, but works exactly the same on all platforms.
> 
> > Then if all apps go into /Application or ~/Application this too could become a standard.
> 
> Well, if we've got desktop2app, who says we can't have an app2desktop?
> 

desktop2app?? app2desktop?? what?? :)

> > If we could then ask all programmers to adhere to certain guidelines, keep the wuality of apps to a certain level, people are willing to pay indie dev’s i am sure of this. appshop could create a backend for items on startup, keep a list of files to be removed when the app is removed. Imagine we go with the hS idea of app bundles.. we could run a daemon that keeps track of files to remove if the programs report these files to the correct daemon? Shouldn’t be too hard to write into a program.
> 
> This is going a bit further than I originally planned, but OK!

Well.. i dare to dream big :) If you dont reach for the stars, you'll never reach the sky :)

**Comment by jhjacobs81** [comment]
i tried GNOME on Debian.. it felt like i'm on Windows, i dont know how to find shit :) 

all these *X problems are related to the lack of standards, something Apple has (well.. they control hardware and software, they have only their own standard, but still..) which is why its becoming a succes. It's why i moved to their platform in the first place.

At the same time, while they scream sky high about privacy, it is becoming a problem. especially with them stonewalling more and more of the platform. For me, this is a big no-no. But there just isnt anything equally beautiful and easy. and working in IT fulltime, i just dont want to spend my evenings tinkering anymore. hS got the general idea right, just not the concept of "user friendlyness" which i still believe doesn't mean has to sacrifice privacy.

**Comment by jhjacobs81** [comment]
alas! we should setup a discord place or something.. this is becoming a slowchat :)

**Comment by jhjacobs81** [comment]
![image](https://user-images.githubusercontent.com/5583803/119474075-1573b200-bd4c-11eb-812a-21a241232150.png)

I do sort of like this layout :)

**Comment by kettle-7** [comment]
I agree. What I don't like is the 15 minutes it takes to get there.

About Debian GNOME, there is a website called extensions.gnome.org where you can get some extensions to make it easier to use. It's best if you install the Firefox extension so it can install them from the website.

I recommend dash to dock and desktop icons.

**Comment by kettle-7** [comment]
> hS got the general idea right, just not the concept of "user friendlyness" which i still believe doesn't mean has to sacrifice privacy.

I agree, but everybody has different morals. Who knows where we'll be in a month!

**Comment by jhjacobs81** [comment]
Get where??

**Comment by kettle-7** [comment]
Get which where?

**Comment by jhjacobs81** [comment]
Haha you conduse me :) 
In any case, if we could mix the microsoft basics with a little dash of Apple and a saus of our own..

**Comment by kettle-7** [comment]
> Eventually, appshop could make sure python is installed, so programmers need not wory about all of this, and just focus on their program?

That would help a lot. Also, if eventually I introduced PyPI/pip integration, that would move a *lot* of people over. Linux and FreeBSD already have relatively up-to-date Python versions installed, macOS falls behind a bit, and Windows doesn't care.

What I need now though is apps to go in the shop, as you can see by the empty shop picture on issue #2.

**Comment by kettle-7** [comment]
> desktop2app?? app2desktop?? what?? :)

Desktop files are the way Linux desktops keep track of apps that are installed, they contain information like the name, description and icon (translated).

There is a tool in helloSystem called desktop2app that takes these legacy files and converts them to .app bundles, which are helloSystem apps.

If you open Filer and in the menu click **Go** > **Go to folder (Alt+Shift+G)** and type /usr/local/share/applications, you can see some of these in action.

If we have a tool to convert desktop files to app bundles, who says we can't do the opposite to convert app bundles to desktop files?

**Comment by jhjacobs81** [comment]
How nice :-) 

i dont have hS anymore, couldn’t get past the 800x600 resolution, useless :( 

would that tool help us to add more apps to AS?

**Comment by kettle-7** [comment]
Yes. It's sort of a one-way thing though -- app2desktop could make a desktop file pointing to an app bundle in /opt, but desktop2app doesn't solve the problem of the app itself still being somewhere else.

Two popular build systems, CMake (which Qt apps mostly use) and Meson (which GNOME apps mostly use) have a very easy way to put the app in a certain place. An even more popular build system (autotools) can't. So it depends on the app. KolourPaint was a nice easy starter.

**Comment by jhjacobs81** [comment]
Would it be hard for you to make such a converter?? See if we can do it ourselves, tell people “hey look! With minimal programming your app gets ready for appshop!” That could convince them to use AS ;-)

**Comment by kettle-7** [comment]
> Would it be hard for you to make such a converter?? See if we can do it ourselves, tell people “hey look! With minimal programming your app gets ready for appshop!” That could convince them to use AS ;-)

Not really. You'd just need information such as icon (which AS handles), description (which AS handles), name (which is easy) and Categories (which is a little harder but that could be included in the metadata).

And as for the app bundles themselves, they're near-universal. I made a script to package a multi-arch AppImage.

**Comment by jhjacobs81** [comment]
Could all the metadata be handled through an xml file? Preferably one that includes the link to the github package directly?? That would allow easily indexing for AS?

**Comment by kettle-7** [comment]
> Could all the metadata be handled through an xml file? Preferably one that includes the link to the github package directly?? That would allow easily indexing for AS?

I've got something simpler than XML at the moment, it's a file called `ls` in each folder of the repository. It points to directories (`d`) and apps (`a`) and it can produce JSON data, which is like XML's successor.

**Comment by jhjacobs81** [comment]
Works for me too.. as long as its easy for developers and us alike ;-)

**Comment by kettle-7** [comment]
Well, it's a work in progress.

The firefox source archive is a lot bigger than I thought!

**Comment by jhjacobs81** [comment]
as is the backend, the website, etc etc ;-) no worries!

**Comment by kettle-7** [comment]
I was just thinking of one of the very few linux distros I can think of that doesn't look like a cartoon - Ubuntu MATE. Debian MATE is probably as good too. I do like Ubuntu MATE's feature where you can change the icon colour, I picked Aqua. The default green gets a little sickly.

**Comment by jhjacobs81** [comment]
🙈🙈 well there’s a lot of work to do still in Linux land :)

**Comment by ghost** [comment]
I don't get the problem,
If developers package their apps as appimages,
Can't we just extract info from the appimage?
